sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/MapperSpec.kt

```diff
@@ -19,158 +19,196 @@ import com.squareup.javapoet.ArrayTypeName
+    if (value.columnDef?.isHandledType ?: true) {
+      code.add(handledTypeGetter(javaType, index, value.element))
+    } else {
+      val factoryField = "${tableInterface!!.simpleName().decapitalize()}$FACTORY_NAME"
+      code.add("$factoryField.${value.columnDef!!.adapterField(nameAllocator)}.$MAP_FUNCTION($CURSOR_PARAM, $index)")
     }
-    if (javaType == TypeName.DOUBLE || javaType == TypeName.DOUBLE.box()) {
-      return CodeBlock.builder().add("$CURSOR_PARAM.getDouble($getter)").build()
+    return code.build()
+  }
+
+  private fun SqliteParser.Column_defContext.cursorGetter(index: Int, interfaceClass: ClassName): CodeBlock {
+    val code = CodeBlock.builder()
+    if (isNullable) {
+      code.add("$CURSOR_PARAM.isNull($index) ? null : ")
     }
-    if (javaType == ArrayTypeName.of(TypeName.BYTE)) {
-      return CodeBlock.builder().add("$CURSOR_PARAM.getBlob($getter)").build()
+    if (isHandledType) {
+      code.add(handledTypeGetter(javaType, index, this))
+    } else {
+      val factoryField = "${interfaceClass.simpleName().decapitalize()}$FACTORY_NAME"
+      code.add("$factoryField.${adapterField(nameAllocator)}.$MAP_FUNCTION($CURSOR_PARAM, $index)")
     }
-    if (javaType == ClassName.get(String::class.java)) {
-      return CodeBlock.builder().add("$CURSOR_PARAM.getString($getter)").build()
+    return code.build()
+  }
+
+  private fun QueryResults.QueryTable.cursorGetter(): CodeBlock {
+    val factoryField = "${interfaceType.simpleName().decapitalize()}$FACTORY_NAME"
+
+    val code = CodeBlock.builder().add("$factoryField.${Table.CREATOR_FIELD}.create(\n")
+    indexedValues.sortedBy { it.index }.map { it.cursorGetter() }.forEachIndexed { i, codeBlock ->
+      if (i != 0) code.add(",\n")
+      code.add("  ").add(codeBlock)
     }
-    throw SqlitePluginException(this, "Unknown cursor getter for type $javaType")
+
+    return code.add("\n)").build()
   }
 
   companion object {
-    private val CREATOR_FIELD = "creator"
-    private val CREATOR_METHOD_NAME = "create"
-    private val CURSOR_TYPE = ClassName.get("android.database", "Cursor")
-    private val CURSOR_PARAM = "cursor"
-    private val MAPPER_TYPE = ClassName.get("com.squareup.sqldelight", "RowMapper")
-    private val MAP_FUNCTION = "map"
     private val NONNULL_TYPE = ClassName.get("android.support.annotation", "NonNull")
+    internal val CURSOR_TYPE = ClassName.get("android.database", "Cursor")
+    internal val CURSOR_PARAM = "cursor"
+    internal val MAPPER_NAME = "Mapper"
+    internal val MAPPER_TYPE = ClassName.get("com.squareup.sqldelight", "RowMapper")
+    internal val MAP_FUNCTION = "map"
+
+    internal fun handledTypeGetter(
+        javaType: TypeName,
+        index: Int,
+        element: ParserRuleContext
+    ): CodeBlock {
+      if (javaType == TypeName.BOOLEAN || javaType == TypeName.BOOLEAN.box()) {
+        return CodeBlock.builder().add("$CURSOR_PARAM.getInt($index) == 1").build()
```

Comment:
`CodeBlock.of` would be perfect here but we can't use it because it's in a too-new JavaPoet.